packages/roosterjs-editor-plugins/lib/PasteManager/PasteManager.ts

```diff
@@ -59,129 +73,77 @@ export default class PasteManager implements EditorPlugin {
-                insertOnNewLine: false,
-            });
+    private pasteHTML(clipboardData: ClipBoardData, fragmentOverride?: DocumentFragment) {
+        if (clipboardData.textData || !clipboardData.imageData.file) {
+            let fragment = fragmentOverride || clipboardData.htmlFragment;
+            convertPastedContentFromWord(fragment);
+            this.editor.insertNode(fragment);
         } else {
-            this.pasteContainer.style.display = '';
+            let pasteHandler = this.imageHandler || this.defaultImageHandler;
+            pasteHandler(clipboardData);
         }
-        this.pasteContainer.focus();
-
-        window.requestAnimationFrame(() => {
-            // restore original selection range in editor
-            editor.updateSelection(originalSelectionRange);
-            callback(this.pasteContainer);
-            this.pasteContainer.style.display = 'none';
-            this.pasteContainer.innerHTML = '';
-        });
-    }
-}
 
-function getClipboardData(dataTransfer: DataTransfer): ClipBoardData {
-    let clipboardData: ClipBoardData = {
-        imageData: {},
-        textData: dataTransfer.getData('text'),
-        htmlData: null,
-    };
-
-    let image = getImage(dataTransfer);
-    if (image) {
-        clipboardData.imageData.file = image;
     }
 
-    return clipboardData;
-}
-
-function getImage(dataTransfer: DataTransfer): File {
-    // Chrome, Firefox, Edge support dataTransfer.items
-    let fileCount = dataTransfer.items ? dataTransfer.items.length : 0;
-    for (let i = 0; i < fileCount; i++) {
-        let item = dataTransfer.items[i];
-        if (item.type && item.type.indexOf('image/') == 0) {
-            return item.getAsFile();
-        }
-    }
-    // IE, Safari support dataTransfer.files
-    fileCount = dataTransfer.files ? dataTransfer.files.length : 0;
-    for (let i = 0; i < fileCount; i++) {
-        let file = dataTransfer.files.item(i);
-        if (file.type && file.type.indexOf('image/') == 0) {
-            return file;
-        }
+    private pasteText(clipboardData: ClipBoardData) {
+        let text = clipboardData.textData || '';
+        let html = text.replace(/\n/g, '<br>').replace(/\s/g, '&nbsp;');
+        this.editor.insertContent(html);
     }
-    return null;
-}
-
-function normalizeContent(content: string): string {
-    // Remove 'position' style from source HTML
-    content = content.replace(INLINE_POSITION_STYLE, '$1');
 
-    // Replace <BR> with <DIV>
-    if (TEXT_WITH_BR_ONLY.test(content)) {
-        content = '<div>' + content.replace(/<br>/gi, '</div><div>') + '<br></div>';
+    private pasteWithCurrentFormat(clipboardData: ClipBoardData) {
+        throw new Error('Not implemented');
     }
 
-    return content;
+    private defaultImageHandler = (clipboardData: ClipBoardData) => {
+        let file = clipboardData.imageData ? clipboardData.imageData.file : null;
+        if (file) {
+            insertImage(this.editor, file);
```

Comment:
In processImages I saw images are categorized into clipboardData.imageData.noSrcImageIds, clipboardData.imageData.localFileImages and clipboardData.imageData.base64Images, are they treated the same way here?

Reply:
That is to give imageHandler a chance to deal with those images (e.g., remove noSrcImage and localFileImages).